#########################################################
##########Convergent Genomics for PCa####################
#########################################################

##RScript Generated by Benjamin B. Currall
##Last Modified 08/20/15


#########################################################
##########1.0 Prepare Environment########################
#########################################################

########################################################
###1.1 Install Packages
#1.1 install packages
source("http://bioconductor.org/biocLite.R")
biocLite(c("org.Hs.eg.db"))
biocLite("gwascat")
install.packages(c("XML"))
install.packages("RCurl")
install.packages("rjson")
install.packages("SSOAP")
install.packages("venneuler")
install.packages("rJava")
install.packages('VennDiagram')
install.packages('plyr')
install.packages('cgdsr')

Sys.setenv(JAVA_HOME='C:\\Program Files\\Java\\jre1.8.0_60')
########################################################
###1.2 Load libraries
library(XML)
library(RCurl)
library(org.Hs.eg.db)
library(gwascat)
library(limma)
library (rjson)
library(reshape2)
library(rJava)
library(venneuler)
library(VennDiagram)
library(plyr)
library(cgdsr)

########################################################
###1.3 set environemental working drive
user<-Sys.info()["user"][[1]]
wfolder<-"CG"
wdrive<-paste0("C:/Users/",user,"/Desktop/",wfolder)
setwd(wdrive)
wd <- getwd()
ddir <- file.path(wd,"data")
dir.create(ddir)
setwd(ddir)


########################################################
###1.4 Define environemental functions and values
NumOfRef <- function(elem) {
  elem2 <- gsub(",","",elem)
  return (nchar(elem) - nchar(elem2)+1)
}

EmpGeneDF<-data.frame(Cat=0,SubCat1 = 0,SubCat2=0,SubCat3=0,GeneSymUsed=0,NCBI_ID=0, NCBI_Sym=0, Ref = 0)

MyID<-"###"
MyPass<-"###"

#########################################################
##########2.0 Retrieve Germline Genes####################
#########################################################


#########################################################
###2.1 Retrieve XML from NCI for Inherited PCa Genes

##download xml
##see http://www.cancer.gov/types/prostate/hp/prostate-genetics-pdq/#section/_981
url<-"http://www.cancer.gov/types/prostate/hp/prostate-genetics-pdq"
ihg.xml <- getURL(url)

##parse xml
ihg.parsed<-htmlTreeParse(ihg.xml, asText = TRUE, useInternal = TRUE)

#(del) ihg.parsed2<-htmlTreeParse(ihg.xml, useInternal = TRUE)
top<-xmlRoot(ihg.parsed)
body<-top[[2]]

##retrieve references
references_xml<-ihg.parsed["//ol"][[4]]
ref_df<-data.frame(num = 0, ref = 0, url = 0)
for(i in 1:(xmlSize(references_xml)/2))
{
  j = ((i*2)-1)
  ref<-xmlValue(references_xml[[j]])
  url<-xmlGetAttr(references_xml[[j]][[2]], "href")
  ref_df[i,]<-c(i,ref,url)
}


##parse Tables
tables = readHTMLTable(ihg.xml)

tables_list<-list()
#tables[1]: not used
#tables[7]: not used
#tables[8]: not used
#tables[9]: not used
#tables[10]: not used
#tables[11]: not used
#tables[12]: not used
#tables[13]: not used
#tables[14]: not used
#tables[15]: not used
#tables[16]: not used
#tables[17]: not used



##tables[2], Table 2. Proposed Prostate Cancer Susceptibility Loci
table2<-data.frame(tables[2])

genes_table2<-EmpGeneDF

Cat<-"Genetic"
SubCat1<-"Germline"
SubCat2<-"Mndl"
SubCat3<-""

for(i in 1:length(which(grepl("OMIM",table2$X_1078.Gene))))
{
  j<-which(grepl("OMIM",table2$X_1078.Gene))[i]
  GeneSymRaw<-as.character(table2$X_1078.Gene[j])
  GeneSymFirst<-gsub("\n.*","", GeneSymRaw)
  if(length(alias2Symbol(GeneSymFirst))>0)
  {
    GeneSymUsed<-GeneSymFirst
    NCBI_Sym <- alias2Symbol(GeneSymFirst)
    NCBI_ID <-unlist(mget(NCBI_Sym,org.Hs.egALIAS2EG))
    
  }else
  {
    duptest<-strsplit(GeneSymRaw,"\\(")
    if(length(grep("OMIM",duptest[[1]]))>1)  
    {
      GeneSymSecond_T<-gsub(".*/","", GeneSymRaw)
      GeneSymSecond<-gsub(" .*","", GeneSymSecond_T)
      if(length(alias2Symbol(GeneSymSecond))>0)
      {
        GeneSymUsed<-GeneSymSecond
        NCBI_Sym <- alias2Symbol(GeneSymSecond)
        NCBI_ID <-unlist(mget(NCBI_Sym,org.Hs.egALIAS2EG))
      }else
      {
        GeneSymUsed<-GeneSymFirst
        NCBI_ID <-"unknown"
        NCBI_Sym <- "unknown"
      }
    }
  }
  
  Ref_t<-gsub(".*\\[","", GeneSymRaw)
  Ref<-gsub("\\].*","", Ref_t)
  
  genes_table2[i,]<-c(Cat,SubCat1,SubCat2,SubCat3,GeneSymUsed,NCBI_ID[1],NCBI_Sym[1],Ref)
}
listlen<-length(tables_list)
tables_list[[listlen+1]]<-genes_table2

##tables[3], Table 3. Case-Control Studies in Genes With Some Association With Prostate Cancer Risk

genes_table3<-EmpGeneDF

table3<-data.frame(tables[3])

Cat<-"Genetic"
SubCat1<-"Germline"
SubCat2<-"CC"
SubCat3<-""
for(i in 1:length(which(grepl("OMIM",table3$X_1159.Gene))))
{
  j<-which(grepl("OMIM",table3$X_1159.Gene))[i]
  GeneSymRaw<-as.character(table3$X_1159.Gene[j])
  GeneSymUsed<-gsub("\n.*)","", GeneSymRaw)
  GeneSymUsed<-gsub(" \n.*)","", GeneSymUsed)
  GeneSymUsed<-gsub(" \\(OMIM\\)","", GeneSymUsed)
  if(length(alias2Symbol(GeneSymUsed))>0)
  {
    NCBI_Sym <- alias2Symbol(GeneSymUsed)
    NCBI_ID <-unlist(mget(NCBI_Sym,org.Hs.egALIAS2EG))
  }else
  {
    NCBI_ID <-"unknown"
    NCBI_Sym <- "unknown"
  }
  
  if((i+1)>length(which(grepl("OMIM",table3$X_1159.Gene))))
    {k<-length(table3$X_1159.Gene)+1}else
    {k <-which(grepl("OMIM",table3$X_1159.Gene))[i+1]}
  Ref<-as.character()
  for(l in (i+1):(k-1))
  {
    RefRaw<-as.character(table3$X_1159.Gene[l])
    if(grepl("\\[",RefRaw))
    {
      Ref_t<-gsub(".*\\[","", RefRaw)
      Ref_t1<-gsub("\\].*","", Ref_t)
      if(length(Ref)<1){Ref<-Ref_t1}else{Ref<-paste0(Ref,",",Ref_t1)}
    }else{Ref<-paste0(Ref)}
  }
  genes_table3[i,]<-c(Cat,SubCat1,SubCat2,SubCat3,GeneSymUsed,NCBI_ID[1],NCBI_Sym[1],Ref)
}
listlen<-length(tables_list)
tables_list[[listlen+1]]<-genes_table3


##tables[4], Table 4. GWAS European
table4<-data.frame(tables[4])
colnames(table4)<-c("SNP","Locus","Gene","Region","Ref", "OR")
table4$position<-as.numeric(as.character(rownames(table4)))

t4_1A<-table4[grepl("[:digit:]",table4$Region),]
t4_1B<-table4[!grepl("[:digit:]",table4$Region),]
t4_1B[,"empty"] <- NA
t4_1C<-data.frame(t4_1B$SNP,t4_1B$empty,t4_1B$Locus,t4_1B$Gene,t4_1B$Region,t4_1B$Ref,t4_1B$position)
colnames(t4_1C)<-colnames(table4)
t4_1Fin<-rbind(t4_1A,t4_1C)
t4_1Fin<-t4_1Fin[order(t4_1Fin$position),]

t4_2A<-t4_1Fin[grepl("[:digit:]",t4_1Fin$Region),]
t4_2B<-t4_1Fin[!grepl("[:digit:]",t4_1Fin$Region),]
t4_2B[,"empty"] <- NA
t4_2C<-data.frame(t4_2B$SNP,t4_2B$empty,t4_2B$Locus,t4_2B$Gene,t4_2B$Region,t4_2B$Ref,t4_2B$position)
colnames(t4_2C)<-colnames(table4)
t4_2Fin<-rbind(t4_2A,t4_2C)
t4_2Fin<-t4_2Fin[order(t4_2Fin$position),]

t4_2Fin$Gene<-as.character(t4_2Fin$Gene)
t4_2Fin$Ref<-as.character(t4_2Fin$Ref)
rownames(t4_2Fin)<-t4_2Fin$position

genes_table4<-EmpGeneDF

Cat<-"Genetic"
SubCat1<-"Germline"
SubCat2<-"GWAS"
SubCat3<-"EU"

pos<-1
for(i in 1:length(t4_2Fin$Gene))
{
  j<-i
  if(!is.na(t4_2Fin$Gene[i])&&t4_2Fin$Gene[i]!="None")
  {
   GeneSymUsed<-t4_2Fin$Gene[i]
   if(length(alias2Symbol(GeneSymUsed))>0)
   {
     NCBI_Sym <- alias2Symbol(GeneSymUsed)
     NCBI_ID <-unlist(mget(NCBI_Sym,org.Hs.egALIAS2EG))
   }else
   {
     NCBI_ID <-"unknown"
     NCBI_Sym <- "unknown"
   }
   RefRaw<-t4_2Fin$Ref[i]
   Ref_t<-gsub(".*\\[","", RefRaw)
   Ref<-gsub("\\].*","", Ref_t)
  }else if(!is.na(t4_2Fin$Gene[i]=="None")&&t4_2Fin$Gene[i]=="None")
  {
    GeneSymUsed<-"None"
    NCBI_ID <-NA
    NCBI_Sym <- NA
    RefRaw<-t4_2Fin$Ref[i]
    Ref_t<-gsub(".*\\[","", RefRaw)
    Ref<-gsub("\\].*","", Ref_t)
    
  }else if(is.na(t4_2Fin$Gene[i]))
  {
    while(is.na(t4_2Fin$Gene[j]))
    {
      RefRaw<-t4_2Fin$Ref[j]
      Ref_t<-gsub(".*\\[","", RefRaw)
      Ref_t1<-gsub("\\].*","", Ref_t)
      Ref<-paste0(Ref,",",Ref_t1)
      j<-j+1
    }
    genes_table4[pos-1,]<-c(Cat,SubCat1,SubCat2,SubCat3,GeneSymUsed,NCBI_ID[1],NCBI_Sym[1],Ref)
    next()
  }
  genes_table4[pos,]<-c(Cat,SubCat1,SubCat2,SubCat3,GeneSymUsed,NCBI_ID[1],NCBI_Sym[1],Ref)
  pos <- pos+1
}
listlen<-length(tables_list)
tables_list[[listlen+1]]<-genes_table4

#tables[5], Table 5. GWAS Non-European
table5<-data.frame(tables[5])
colnames(table5)<-c("SNP","Locus","Gene","Region","Ref", "OR")
table5$position<-as.numeric(as.character(rownames(table5)))

t5_1A<-table5[grepl("[:digit:]",table5$Region),]
t5_1B<-table5[!grepl("[:digit:]",table5$Region),]
t5_1B[,"empty"] <- NA
t5_1C<-data.frame(t5_1B$SNP,t5_1B$empty,t5_1B$Locus,t5_1B$Gene,t5_1B$Region,t5_1B$Ref,t5_1B$position)
colnames(t5_1C)<-colnames(table5)
t5_1Fin<-rbind(t5_1A,t5_1C)
t5_1Fin<-t5_1Fin[order(t5_1Fin$position),]

t5_2A<-t5_1Fin[grepl("[:digit:]",t5_1Fin$Region),]
t5_2B<-t5_1Fin[!grepl("[:digit:]",t5_1Fin$Region),]
t5_2B[,"empty"] <- NA
t5_2C<-data.frame(t5_2B$SNP,t5_2B$empty,t5_2B$Locus,t5_2B$Gene,t5_2B$Region,t5_2B$Ref,t5_2B$position)
colnames(t5_2C)<-colnames(table5)
t5_2Fin<-rbind(t5_2A,t5_2C)
t5_2Fin<-t5_2Fin[order(t5_2Fin$position),]

t5_2Fin$Gene<-as.character(t5_2Fin$Gene)
t5_2Fin$Ref<-as.character(t5_2Fin$Ref)
rownames(t5_2Fin)<-t5_2Fin$position

genes_table5<-EmpGeneDF

Cat<-"Genetic"
SubCat1 <- "Germline"
SubCat2 <- "GWAS"
SubCat3 <- ""

pos<-1
for(i in 1:length(t5_2Fin$Gene))
{
  j<-i
  if(is.na(t5_2Fin$Locus[i]))
  {
    if(t5_2Fin$SNP[i]=="African American Population"){SubCat3<-"AA"}else
    if(t5_2Fin$SNP[i]=="Chinese Population"){SubCat3<-"AP"}else
    if(t5_2Fin$SNP[i]=="Japanese Population"){SubCat3<-"JPN"}
    next()
  }else if(!is.na(t5_2Fin$Gene[i])&&t5_2Fin$Gene[i]!="None")
  {
    GeneSymUsed<-t5_2Fin$Gene[i]
    if(length(alias2Symbol(GeneSymUsed))>0)
    {
      NCBI_Sym <- alias2Symbol(GeneSymUsed)
      NCBI_ID <-unlist(mget(NCBI_Sym,org.Hs.egALIAS2EG))
    }else
    {
      NCBI_ID <-"unknown"
      NCBI_Sym <- "unknown"
    }
    RefRaw<-t5_2Fin$Ref[i]
    Ref_t<-gsub(".*\\[","", RefRaw)
    Ref<-gsub("\\].*","", Ref_t)
  }else if(!is.na(t5_2Fin$Gene[i]=="None")&&t5_2Fin$Gene[i]=="None")
  {
    GeneSymUsed<-"None"
    NCBI_ID <-NA
    NCBI_Sym <- NA
    RefRaw<-t5_2Fin$Ref[i]
    Ref_t<-gsub(".*\\[","", RefRaw)
    Ref<-gsub("\\].*","", Ref_t)
    
  }else if(is.na(t5_2Fin$Gene[i]))
  {
    while(is.na(t5_2Fin$Gene[j]))
    {
      RefRaw<-t5_2Fin$Ref[j]
      Ref_t<-gsub(".*\\[","", RefRaw)
      Ref_t1<-gsub("\\].*","", Ref_t)
      Ref<-paste0(Ref,",",Ref_t1)
      j<-j+1
    }
    genes_table5[pos-1,]<-c(Cat,SubCat1,SubCat2,SubCat3,GeneSymUsed,NCBI_ID[1],NCBI_Sym[1],Ref)
    next()
  }
  genes_table5[pos,]<-c(Cat,SubCat1,SubCat2,SubCat3,GeneSymUsed,NCBI_ID[1],NCBI_Sym[1],Ref)
  pos <- pos+1
}
listlen<-length(tables_list)
tables_list[[listlen+1]]<-genes_table5

#tables[6], Table 6. Aggressive Inherited Variants
genes_table6<-EmpGeneDF

table6<-data.frame(tables[6])
colnames(table6)<-c("Gene","Variant","Phenotype","Controls","OR", "Ref")
table6$position<-as.numeric(as.character(rownames(table6)))
t6<-table6[grepl("Target gene", table6$Gene),]
t6$Gene<-gsub(".* ", "",t6$Gene)
t6$Ref<-gsub(c("[[:punct:]]"),"", t6$Ref)
t6$NCBI.Sym<-alias2SymbolTable(t6$Gene)

Cat<-"Genetic"
SubCat1<-"Germline"
SubCat2<-"NS"
SubCat3<-"Aggressive"

for(i in 1:length(t6$NCBI.Sym))
{
  GeneSymUsed<-t6$NCBI.Sym[i]
  if(length(alias2Symbol(GeneSymUsed))>0)
  {
    NCBI_Sym <- alias2Symbol(GeneSymUsed)
    NCBI_ID <-unlist(mget(NCBI_Sym,org.Hs.egALIAS2EG))
  }else
  {
    NCBI_ID <-"unknown"
    NCBI_Sym <- "unknown"
  }
  Ref<-t6$Ref[i]
  
  genes_table6[i,]<-c(Cat,SubCat1,SubCat2,SubCat3,GeneSymUsed,NCBI_ID[1],NCBI_Sym[1],Ref)
}
listlen<-length(tables_list)
tables_list[[listlen+1]]<-genes_table6

##Full NCI Table

NCI_genes<-do.call(rbind, tables_list)


###################################################################################
###2.2 Retrieve ftp from OMIM for Inherited PCa Genes
tables_list<-list()

##download xml
ftp<-"ftp://ftp.omim.org/OMIM/morbidmap"
download.file(ftp,"omim_morbid.txt")
OMIM_table<-read.table("omim_morbid.txt", sep = "|", col.names = c("Disease","Gene","OMIM_ID","Locus"), quote = "")

OMIM_PCa<-OMIM_table[grepl("Prostate",OMIM_table$Disease),]

###OMIM Inherited
OMIM_PCa_Inherited<-OMIM_PCa[!grepl("somatic",OMIM_PCa$Disease),]
OMIM_Inherited<-EmpGeneDF

Cat<-"Genetic"
SubCat1<-"Germline"
SubCat2<-"NS"
SubCat3<-""
Ref<-"OMIM"

pos<-1
for(i in 1:length(OMIM_PCa_Inherited$Disease))
{
  GeneList<-strsplit(as.character(OMIM_PCa_Inherited$Gene[i]), ",")
  GeneList<-gsub(" ", "", GeneList[[1]])
  for(j in 1:length(GeneList))
  {
    GeneSymUsed<-GeneList[j]
    if(length(alias2Symbol(GeneSymUsed))>0)
    {
      NCBI_Sym <- alias2Symbol(GeneSymUsed)
      NCBI_ID <-unlist(mget(NCBI_Sym,org.Hs.egALIAS2EG))
      
    }else
    {
      NCBI_ID <-"unknown"
      NCBI_Sym <- "unknown"
    }
    OMIM_Inherited[pos,]<-c(Cat,SubCat1,SubCat2,SubCat3, GeneSymUsed, NCBI_ID[1], NCBI_Sym[1],Ref)
    pos <-pos+1
  }
}
OMIM_Inherited<-OMIM_Inherited[which(!duplicated(OMIM_Inherited$NCBI_ID)),]

listlen<-length(tables_list)
tables_list[[listlen+1]]<-OMIM_Inherited

###OMIM Somatic
OMIM_PCa_Somatic<-OMIM_PCa[grepl("somatic",OMIM_PCa$Disease),]

OMIM_Somatic<-EmpGeneDF

Cat<-"Genetic"
SubCat1<-"Somatic"
SubCat2<-""
SubCat3<-""
Ref<-"OMIM"

pos<-1
for(i in 1:length(OMIM_PCa_Somatic$Disease))
{
  GeneList<-strsplit(as.character(OMIM_PCa_Somatic$Gene[i]), ",")
  GeneList<-gsub(" ", "", GeneList[[1]])
  for(j in 1:length(GeneList))
  {
    GeneSymUsed<-GeneList[j]
    if(length(alias2Symbol(GeneSymUsed))>0)
    {
      NCBI_Sym <- alias2Symbol(GeneSymUsed)
      NCBI_ID <-unlist(mget(NCBI_Sym,org.Hs.egALIAS2EG))
    
    }else
    {
      NCBI_ID <-"unknown"
      NCBI_Sym <- "unknown"
    }
    OMIM_Somatic[pos,]<-c(Cat, SubCat1, SubCat2,SubCat3, GeneSymUsed, NCBI_ID[1], NCBI_Sym[1],Ref)
    pos <-pos+1
  }
}
OMIM_Somatic<-OMIM_Somatic[which(!duplicated(OMIM_Somatic$NCBI_ID)),]


listlen<-length(tables_list)
tables_list[[listlen+1]]<-OMIM_Somatic

##Full NCI Table

OMIM_genes<-do.call(rbind, tables_list)

###################################################################################
###2.3 GWAS genes

#data(gwrngs19)
#subsetByTraits(gwrngs19, tr="prostate")[1:3]

http<-"http://www.genome.gov/admin/gwascatalog.txt"
download.file(http,"gwascatalog.txt")
gwahits = data.frame(read.table("gwascatalog.txt",header=TRUE,sep='\t',quote="",comment.char="",as.is=TRUE))

gwas_PCa<-gwahits[grepl("Prostate cancer", gwahits$Disease.Trait),]

GWASCat<-EmpGeneDF

Cat<-"Genetic"
SubCat1<-"Germline"
SubCat2<-"GWAS"
SubCat3<-""
pos<-1
for(i in 1:length(gwas_PCa$Reported.Gene.s.))
{
  GeneList<-strsplit(as.character(gwas_PCa$Reported.Gene.s.[i]), ",")
  GeneList<-gsub(" ", "", GeneList[[1]])
  for(j in 1:length(GeneList))
  {
    GeneSymUsed<-GeneList[j]
    if(length(alias2Symbol(GeneSymUsed))>0)
    {
      NCBI_Sym <- alias2Symbol(GeneSymUsed)
      NCBI_ID <-unlist(mget(NCBI_Sym,org.Hs.egALIAS2EG))
      
    }else
    {
      NCBI_ID <-"unknown"
      NCBI_Sym <- "unknown"
    }
    Ref<-gwas_PCa$PUBMEDID[i]
    GWASCat[pos,]<-c(Cat, SubCat1, SubCat2, SubCat3, GeneSymUsed, NCBI_ID[1], NCBI_Sym[1],Ref)
    pos <-pos+1
  }
}

GWASCat2<-GWASCat[!grepl("unknown", GWASCat$NCBI_ID),]
GWASCat2a<-GWASCat2[which(!duplicated(GWASCat2[,c("NCBI_ID","Ref")])),]
GWASCat2a<-aggregate(GWASCat2[-6], by=list(GWASCat2$NCBI_ID),c)
colnames(GWASCat2a)[1]<-"NCBI_ID"
GWASCat2a$Cat<-"Genetic"
GWASCat2a$SubCat1<-"Germline"
GWASCat2a$SubCat2<-"GWAS"
GWASCat2a$SubCat3<-""
GWASCat2a$GeneSymUsed<-lapply(GWASCat2a$GeneSymUsed,function(elm) paste(elm[which(!duplicated(elm))], collapse = ","))
GWASCat2a$NCBI_Sym<-lapply(GWASCat2a$NCBI_Sym, function(elm) elm[[1]])
GWASCat2a$Ref<-lapply(GWASCat2a$Ref,function(elm) paste(elm[which(!duplicated(elm))], collapse = ","))
GWASCat2a<-GWASCat2a[,c("Cat","SubCat1","SubCat2","SubCat3","GeneSymUsed","NCBI_ID","NCBI_Sym","Ref")]


GWASCat3<-GWASCat[grepl("unknown", GWASCat$NCBI_ID),]
GWASCat3a<-GWASCat3[!grepl("NR", GWASCat3$GeneSymUsed),]
GWASCat3a<-GWASCat3a[!grepl("Intergenic", GWASCat3a$GeneSymUsed),]
GWASCat3a<-aggregate(GWASCat3a[-5], by=list(GWASCat3a$GeneSymUsed),c)
colnames(GWASCat3a)[1]<-"GeneSymUsed"
GWASCat3a$Cat<-"Genetic"
GWASCat3a$SubCat1<-"Germline"
GWASCat3a$SubCat2<-"GWAS"
GWASCat3a$SubCat3<-""
GWASCat3a$NCBI_ID<-lapply(GWASCat3a$NCBI_ID,function(elm) elm[[1]])
GWASCat3a$NCBI_Sym<-lapply(GWASCat3a$NCBI_Sym, function(elm) elm[[1]])
GWASCat3a$Ref<-lapply(GWASCat3a$Ref,function(elm) paste(elm[which(!duplicated(elm))], collapse = ","))
GWASCat3a<-GWASCat3a[,c("Cat","SubCat1","SubCat2","SubCat3","GeneSymUsed","NCBI_ID","NCBI_Sym","Ref")]


GWASCatf<-rbind(GWASCat2a,GWASCat3a)


###############################################
#Somatic Genes
# url<-c("sftp-cancer.sanger.ac.uk/files/grch38/cosmic/v74/")
# x<-getURL(url, userpwd= paste0(MyID,":",MyPass))
# url<-"http://cancer.sanger.ac.uk/cosmic/census/all?name=all&home=y&sEcho=1&export=csv"
# download.file(url,"CosmicCon.txt")

Cosmic <- read.csv("Census_all.csv")
CosmicPCa<-Cosmic[grep("prostate",Cosmic$Tumour.Types.Somatic.),]
CosmicPCa<-CosmicPCa[,c("Entrez.GeneId","Gene.Symbol","Somatic","Germline")]
colnames(CosmicPCa)<-c("NCBI_ID","GeneSymUsed","Somatic","Germline")
NCBI_ID<-as.character(CosmicPCa$NCBI_ID)
NCBI_ID <- NCBI_ID[NCBI_ID %in% ls(org.Hs.egSYMBOL)]
NCBI_Sym <-unlist(mget(NCBI_ID,org.Hs.egSYMBOL))
genenames_NCBI<-data.frame(NCBI_ID,NCBI_Sym)
genenames_NCBI<-genenames_NCBI[which(!duplicated(genenames_NCBI$NCBI_Sym)),]
CosmicPCa <- merge(CosmicPCa, genenames_NCBI, by="NCBI_ID", all.x = TRUE)

CosmicGerm<-CosmicPCa[grep("yes",CosmicPCa$Germline),]
CosmicGerm$Cat<-"Genetic"
CosmicGerm$SubCat1<-"Germline"
CosmicGerm$SubCat2<-"NS"
CosmicGerm$SubCat3<-""
CosmicGerm$Ref<-"COSMIC"

CosmicSom<-CosmicPCa[grep("yes",CosmicPCa$Somatic),]
CosmicSom$Cat<-"Genetic"
CosmicSom$SubCat1<-"Somatic"
CosmicSom$SubCat2<-""
CosmicSom$SubCat3<-""
CosmicSom$Ref<-"COSMIC"

CosmicGenes<-rbind(CosmicGerm,CosmicSom)

CosmicGenes<-CosmicGenes[,c("Cat","SubCat1","SubCat2","SubCat3","GeneSymUsed","NCBI_ID","NCBI_Sym","Ref")]


##TCGA Somatic Mutations
#copied manually from: http://www.cbioportal.org/study.do?cancer_study_id=prad_tcga#mutations on 10/1/15
#saved in TcgaMS.csv
TcgaMS<-read.csv("TcgaMS.csv")
TcgaMS<-TcgaMS[which(!is.na(TcgaMS$MutSig)),]

GeneConv <- select(org.Hs.eg.db, keys=c(as.character(TcgaMS$Gene)), columns=c("SYMBOL","ENTREZID"), keytype="SYMBOL")
TcgaMS$NCBI_ID<-GeneConv$ENTREZID
colnames(TcgaMS)[1]<-"GeneSymUsed"
TcgaMS$Cat<-"Genetic"
TcgaMS$SubCat1<-"Somatic"
TcgaMS$SubCat2<-""
TcgaMS$SubCat3<-""
TcgaMS$NCBI_Sym<-""
TcgaMS$Ref<-"TCGA"
TCGAGenes<-TcgaMS[,c("Cat","SubCat1","SubCat2","SubCat3","GeneSymUsed","NCBI_ID","NCBI_Sym","Ref")]

###############################################
###My Genes Germline
MyGenesG<-EmpGeneDF
MyGenesG$GeneSymUsed[1]<-"C2orf43"
MyGenesG$NCBI_ID<-NULL

GeneConv <- select(org.Hs.eg.db, keys=c(as.character(MyGenesG$GeneSymUsed)), columns=c("SYMBOL","ENTREZID"), keytype="SYMBOL")
colnames(GeneConv)<-c("GeneSymUsed","NCBI_ID")
MyGenesG<-join(MyGenesG,GeneConv, by="GeneSymUsed", type = "left")
MyGenesG$Cat<-"Genetic"
MyGenesG$SubCat1<-"Germline"
MyGenesG$SubCat2<-"DeNovo"
MyGenesG$SubCat3<-"Translocation"
MyGenesG$NCBI_Sym<-""
MyGenesG$Ref<-"CS"
MGGenes<-MyGenesG[,c("Cat","SubCat1","SubCat2","SubCat3","GeneSymUsed","NCBI_ID","NCBI_Sym","Ref")]



###################################################################################
##################3.0 Expression Data##############################################
###################################################################################

###################################################################################
###3.1 Transcription genes
## Derived from microarray studies

PCa_RO<-read.table("alldata_fullstudy_ro_plot.txt")
PCa_RO$NCBI_Sym<-rownames(PCa_RO)
ID.r <- unique(as.vector(as.matrix(PCa_RO$NCBI_Sym)))
ID.r <- ID.r[ID.r %in% ls(org.Hs.egALIAS2EG)]
NCBI_ID <-unlist(mget(ID.r,org.Hs.egALIAS2EG))
NCBI_ID<-unique(NCBI_ID)
NCBI_Sym <- unlist(mget(NCBI_ID,org.Hs.egSYMBOL))
genenames_NCBI<-data.frame(NCBI_ID,NCBI_Sym)
genenames_NCBI<-genenames_NCBI[which(!duplicated(genenames_NCBI$NCBI_Sym)),]
PCa_RO_IDs <- merge(PCa_RO, genenames_NCBI, by="NCBI_Sym", all.x = TRUE)

PCaSigExp_H<-PCa_RO_IDs[which(PCa_RO_IDs$ro_t>=0.95),]
PCaSigExp_H$SubCat1<-"OG"
PCaSigExp_L<-PCa_RO_IDs[which(PCa_RO_IDs$ro_t<=0.05),]
PCaSigExp_L$SubCat1<-"TS"

PCaSigExp<-rbind(PCaSigExp_H,PCaSigExp_L)

PCaSigExp$Cat<-"Exp"
PCaSigExp$Ref<-"CS"
PCaSigExp$GeneSymUsed<-""

Exp_genes<-PCaSigExp[,c("Cat","SubCat1","GeneSymUsed","NCBI_ID", "NCBI_Sym","Ref")]



###################################################################################
##################4.0 Pathway Data#################################################
###################################################################################

###################################################################################
###4.1 Pathways
## Derived from wikipaths and KEGG

##wikipathways
# pwId<-"WP2263"
# code<-"L" #L is NCBI Gene ID
# method<-"getXrefList"
# format<-"json"
# Url<-paste0("http://webservice.wikipathways.org/index.php?pwId=",pwId,"&code=",code,"&method=",method,"&format=",format)
# Json <- getURL(Url)
# Resp <- data.frame(fromJSON(Json))
# colnames(Resp)<-"NCBI_ID"
# 


Resp<-read.delim("WP2263_80439.txt", stringsAsFactors = FALSE)
colnames(Resp)<-c("NCBI_ID","db")
Resp<-Resp[grep("Entrez", Resp$db),]
Resp<-Resp[which(!duplicated(Resp$NCBI_ID)),]

NCBI_ID<-as.character(Resp$NCBI_ID)
NCBI_ID <- NCBI_ID[NCBI_ID %in% ls(org.Hs.egSYMBOL)]
NCBI_Sym <-unlist(mget(NCBI_ID,org.Hs.egSYMBOL))
genenames_NCBI<-data.frame(NCBI_ID,NCBI_Sym)
genenames_NCBI<-genenames_NCBI[which(!duplicated(genenames_NCBI$NCBI_Sym)),]
WPdf <- merge(Resp, genenames_NCBI, by="NCBI_ID", all.x = TRUE)
WPdf$Cat<-"Pathways"
WPdf$Ref<-"WikiPath"
WPdf$db<-NULL

##KEGG Pathways
KGPwId<-"hsa05215" #prostate cancer is "hsa05215"
format<-"json"
Url<-paste0("http://togows.dbcls.jp/entry/pathway/",KGPwId,"/genes.",format)
Json <- getURL(Url)
Resp <- fromJSON(Json)
Resp<-data.frame(names(Resp[[1]]))
colnames(Resp)<-"NCBI_ID"
NCBI_ID<-as.character(Resp$NCBI_ID)
NCBI_ID <- NCBI_ID[NCBI_ID %in% ls(org.Hs.egSYMBOL)]
NCBI_Sym <-unlist(mget(NCBI_ID,org.Hs.egSYMBOL))
genenames_NCBI<-data.frame(NCBI_ID,NCBI_Sym)
genenames_NCBI<-genenames_NCBI[which(!duplicated(genenames_NCBI$NCBI_Sym)),]
KGdf <- merge(Resp, genenames_NCBI, by="NCBI_ID", all.x = TRUE)
KGdf$Cat<-"Pathways"
KGdf$Ref<-"KEGG"

PCaPW<-rbind(WPdf,KGdf)
PCaPW<-PCaPW[,c("Cat","NCBI_ID", "NCBI_Sym","Ref")]

PW_genes<-dcast(PCaPW, Cat+NCBI_ID+NCBI_Sym~Ref)
PW_genes$WikiPath[is.na(PW_genes$WikiPath)]<-""
PW_genes$WikiPath[!is.na(PW_genes$KEGG)]<-paste(PW_genes$WikiPath[!is.na(PW_genes$KEGG)],PW_genes$KEGG[!is.na(PW_genes$KEGG)],sep = ",")
PW_genes$WikiPath<-gsub("^,","",PW_genes$WikiPath)
PW_genes$KEGG<-NULL
colnames(PW_genes)[4]<-"Ref"

###################################################################################
##################5.0 Mouse Model Data#############################################
###################################################################################

###################################################################################

PCaMM<- read.csv("PCaMM.csv")
PCaMM<-PCaMM[,c("Gene","Type", "Hyperplasia..LGPIN","Ref")]
colnames(PCaMM)<-c("NCBI_Sym","SubCat1","LGPIN","Ref")
PCaMM<-PCaMM[which(!grepl("No",PCaMM$LGPIN)),]
PCaMM<-PCaMM[which(!grepl("\\(",PCaMM$NCBI_Sym)),]
PCaMM<-PCaMM[which(PCaMM$NCBI_Sym!="-"),]
PCaMM<-PCaMM[which(!duplicated(PCaMM$NCBI_Sym)),]
PCaMM$NCBI_Sym<-gsub("LDAH","C2orf43",PCaMM$NCBI_Sym)

GeneConv <- select(org.Hs.eg.db, keys=c(as.character(PCaMM$NCBI_Sym)), columns=c("SYMBOL", "ENTREZID"), keytype="SYMBOL")
colnames(GeneConv)<-c("NCBI_Sym","NCBI_ID")
PCaMM <- join(PCaMM, GeneConv, by = "NCBI_Sym", type = "left")

PCaMM$Cat<-"MM"

PCaMM<-PCaMM[,c("Cat","SubCat1","NCBI_ID", "NCBI_Sym","Ref")]


###################################################################################
##################6.0 Combine Data#################################################
###################################################################################

###################################################################################
###Cominbe datasets
AllGenes<-rbind(NCI_genes,OMIM_genes,GWASCatf,CosmicGenes,TCGAGenes,MGGenes)
AllGenes<-AllGenes[which(!grepl("unknown",AllGenes$NCBI_ID)),]
AllGenes<-AllGenes[which(!is.na(AllGenes$NCBI_ID)),]
AllGenes$NCBI_ID<-as.character(AllGenes$NCBI_ID)
AllGenes$Cat<-as.character(AllGenes$Cat)
AllGenes$Ref<-as.character(AllGenes$Ref)


RefSplt<-sapply(AllGenes$Ref, function(e1)
{
  RefCor<- strsplit(e1,",") 
  namelist<-sapply(RefCor, function(e2)
  {
    RefCorA<-e2[grep("-",e2)]
    namelist<-sapply(RefCorA, function(e3)
    {
      edges<-strsplit(e3[grep("-",e3)],"-")
      newname<-sapply(edges,function(e4)
      {
        paste0(seq(as.numeric(as.character(e4[1])),as.numeric(as.character(e4[2]))),collapse=",")
      })
      namelist<-if(grepl("-",e3)){newname}else{e3}
    })
  })
  names(namelist)<-namelist
})

Enumer<-unlist(RefSplt)
Orig<-names(Enumer)
New<-gsub("\\d*-\\d*","",Orig)
New<-gsub("\\.","",New)
NewRef<-data.frame(cbind(Orig,Enumer,New), stringsAsFactors = FALSE)
NewRef$final<-paste0(NewRef$New,NewRef$Enumer)

AllGenes$Ref[grep("-",AllGenes$Ref)]<-NewRef$final

AllGenes$RefCount<-NumOfRef(AllGenes$Ref)
AllGenes$NCBI_Sym<-as.character(AllGenes$NCBI_Sym)
AllGenesW<-dcast(AllGenes, NCBI_ID~SubCat1+SubCat2, fun.aggregate = sum, value.var = "RefCount")

PW_genesR<-PW_genes[,c("NCBI_ID","Ref")]
PW_genesR$PW<-NumOfRef(PW_genesR$Ref)
PW_genesR$Ref<-NULL
AllGenesF<-join(AllGenesW,PW_genesR, by="NCBI_ID", type = "left")

Exp_genesR<-Exp_genes[,c("SubCat1","NCBI_ID")]
colnames(Exp_genesR)<-c("Exp","NCBI_ID")
AllGenesF<-join(AllGenesF,Exp_genesR, by="NCBI_ID", type = "left")

PCaMMR<-PCaMM[,c("NCBI_ID","SubCat1")]
colnames(PCaMMR)<-c("NCBI_ID","MM")
AllGenesF<-join(AllGenesF,PCaMMR, by="NCBI_ID", type = "left")

GeneConv <- select(org.Hs.eg.db, keys=c(as.character(AllGenesF$NCBI_ID)), columns=c("ENTREZID","SYMBOL"), keytype="ENTREZID")
colnames(GeneConv)<-c("NCBI_ID","NCBI_Sym")
AllGenesF<-join(AllGenesF,GeneConv, by="NCBI_ID", type = "left")

###################################################################################
##################7.0 Data Analysis#################################################
###################################################################################

###################################################################################
###Calculate convergent genomics
##Table
CT<-AllGenesF
CT$GL_NG<-apply(CT[c("Germline_CC","Germline_Mndl","Germline_NS","Germline_DeNovo")],1,sum)

#Calculate CG Score
CT$GL_NGn<-CT$GL_NG
CT$GL_NGn[which(CT$GL_NGn>0)]<-1
CT$GL_Gn<-CT$Germline_GWAS
CT$GL_Gn[which(CT$GL_Gn>0)]<-1
CT$Sn<-CT$Somatic_
CT$Sn[which(CT$Sn>0)]<-1
CT$GEScore<-rowSums(CT[c("GL_NGn","GL_Gn","Sn")])

#Calculate CE Score
CT$Expn<-CT$Exp
CT$Expn[!is.na(CT$Expn)]<- 1
CT$Expn[is.na(CT$Expn)]<- 0
CT$Expn<- as.numeric(CT$Expn)

CT$PWn<-CT$PW
CT$PWn[!is.na(CT$PWn)]<- 1
CT$PWn[is.na(CT$PWn)]<- 0
CT$PWn<- as.numeric(CT$PWn)

CT$MMn<-as.character(CT$MM)
CT$MMn[!is.na(CT$MMn)]<- 1
CT$MMn[is.na(CT$MMn)]<- 0
CT$MMn<- as.numeric(CT$MMn)

CT$CEScore<-rowSums(CT[c("Expn","PWn","MMn")])

CT$FScore<-rowSums(CT[c("GEScore","CEScore")])

CT<-CT[,c("NCBI_ID","NCBI_Sym","Germline_GWAS","GL_NG","Somatic_","Exp","PW","MM","FScore")]
CT$MM<-as.character(CT$MM)
CT<-as.character(CT)
CT[is.na(CT)] <- "-"
CT[CT==0]<-"-"

write.table(CT, "CT.csv")

##Diagrams
#Convergenet Genomics 
CG<-AllGenesF
CG<-CG[,c("NCBI_ID","Germline_GWAS","Germline_CC","Germline_Mndl","Germline_DeNovo","Germline_NS","Somatic_")]
CG$GL_NG<-apply(CG[c("Germline_CC","Germline_Mndl","Germline_NS","Germline_DeNovo")],1,sum)
CG<-CG[,c("NCBI_ID","Germline_GWAS","GL_NG","Somatic_")]
rownames(CG)<-CG$NCBI_ID
CG$NCBI_ID<-NULL
CG[CG>1]<-1
colnames(CG)<-c("G","O","S")
CG$"G&O"<-apply(CG[c("G","O")],1,function(x) ifelse(sum(x)>1,1,0))
CG$"G&S"<-apply(CG[c("G","S")],1,function(x) ifelse(sum(x)>1,1,0))
CG$"O&S"<-apply(CG[c("O","S")],1,function(x) ifelse(sum(x)>1,1,0))
CG$"G&O&S"<-apply(CG[c("G","O","S")],1,function(x) ifelse(sum(x)>2,1,0))
CGs<-colSums(CG)

#plot CG euler
su <- venneuler(CGs)
plot(su)

#plot CG Venn
grid.newpage()
draw.triple.venn(
  area1 = CGs[["G"]], area2 = CGs[["O"]], area3 = CGs[["S"]], 
  n12 = CGs[["G&O"]], n23 = CGs[["O&S"]], n13 = CGs[["G&S"]], n123 = CGs[["G&O&S"]], 
  category = c("Gerline_GWAS", "Germline_NG", "Somatic"), lty = "blank", 
  fill = c("skyblue", "pink1", "mediumorchid"),euler.d = TRUE)

##Convergenet Evidence
CE<-AllGenesF
rownames(CE)<-CE$NCBI_ID
CE$CG<-apply(CE[c("Germline_GWAS","Germline_CC","Germline_Mndl","Germline_NS","Germline_DeNovo","Somatic_")],1,sum)
CE<-CE[,c("CG","Exp","PW","MM")]
CE$MM<-as.numeric(CE$MM)
CE[is.na(CE)] <- 0
CE$Exp<-gsub("TS",1,CE$Exp)
CE$Exp<-gsub("OG",1,CE$Exp)
CE$Exp<-as.numeric(CE$Exp)
CE[CE>1]<-1
CE$"Exp&PW"<-apply(CE[c("Exp","PW")],1,function(x) ifelse(sum(x)>1,1,0))
CE$"Exp&MM"<-apply(CE[c("Exp","MM")],1,function(x) ifelse(sum(x)>1,1,0))
CE$"PW&MM"<-apply(CE[c("PW","MM")],1,function(x) ifelse(sum(x)>1,1,0))
CE$"Exp&PW&MM"<-apply(CE[c("Exp","PW","MM")],1,function(x) ifelse(sum(x)>2,1,0))
CEs<-colSums(CE)

#plot CE euler
su <- venneuler(CEs[-1])
plot(su)

#plot CE venn
grid.newpage()
draw.triple.venn(
  area1 = CEs[["Exp"]], area2 = CEs[["PW"]], area3 = CEs[["MM"]], 
  n12 = CEs[["Exp&PW"]], n23 = CEs[["PW&MM"]], n13 = CEs[["Exp&MM"]], n123 = CEs[["Exp&PW&MM"]], 
  category = c("Expresion", "Pathway", "Mouse Model"), lty = "blank", 
  fill = c("skyblue", "pink1", "mediumorchid"),euler.d = TRUE)

